src/porepy/models/abstract_equations.py

```diff
@@ -1,18 +1,78 @@
-"""Abstract equation classes.
+"""Module containing classes representing generalized equations and variables.
+
+- :class:`EquationMixin`: Base equation class providing the interface for setting
+    equations. This interface is used by the solution strategy mixin.
+- :class:`BalanceEquation`: Intermediate base class for vector and scalar balance
+    equations.
+- :class:`LocalElimination`: Intermediate base class for local equations which
+    represent the elimination of some variable by a correlation/function dependent on
+    other variables.
+- :class:`VariableMixin`: Base class for variables, providing the interface for
+    creating variables. This interface is used by the solution strategy mixin.
 
-Contains:
-    - BalanceEquation: Base class for vector and scalar balance equations.
-    - VariableMixin: Base class for variables.
 """
 
 from __future__ import annotations
 
-from typing import Union, cast
+from functools import cached_property
+from typing import Callable, Sequence, Union, cast
+
+import numpy as np
 
 import porepy as pp
+from porepy.numerics.ad.equation_system import GridEntity
+
+
+class EquationMixin(pp.PorePyModel):
```

Comment:
Yes, I think the abstraction is fine. But is "mixin" necessary in the name of this class (and possibly others)?

Reply:
No issue with renaming it. I followed some (probably not intended) convention I saw in the models

`BounderyConditionMixin` inherited by e.g. `BoundaryConditionsSinglePhaseFlow`
`VariableMixin` inherited by `VariablesSinglePhaseFlow` etc.

So I assumed the base classes all have "mixin" in their name, while the concrete physics do not..

Reply:
I think we need some convention in PorePy on how to name or refer to classes which are mixins in the sense that they provide only overloads. The terminology is getting messy, which is affecting the quality of the docs a bit.

Here is probably not the right place, but there is an occurring issue with mypy when creating "true" mixins which only overload methods: If a super-call is performed to the the overloaded method, and the method is declared in the protocols (as intended by the introduction of protocols), mypy raises safe-super errors since version 3.something ("parent method has trivial body" even if base inherits from `Protocol`).
It seems that the newer version deviates from the intentions behind protocols.